Refactored Tools::getCountry function

michaelKaefer opened the pull request:
Method was not commented or explained, did my best to refactor. Changed method definition -> deprecated

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Method was not commented or explained, did my best to refactor. Changed method definition -> deprecated
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | yes
| Fixed ticket? | -
| How to test?  | no existing test found

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/9060)
<!-- Reviewable:end -->

michaelKaefer committed:
Refactor

Method was not commented or explained, did my best to refactor. Changed method definition -> deprecated

mickaelandrieu reviewed (commented):
looks good but I don't feel confident, let's wait for one more review :)

Quetzacoalt91 commented:
When we refacto methods and classes, the previous behavior should be covered by unit tests. Then these tests will confirm nothing has changed between the old and the new version.

I'd be confident in approving this PR if the tests were provided at the same time.

Quetzacoalt91 commented on classes/Tools.php:
```diff
@@ -611,22 +611,58 @@ public static function switchLanguage(Context $context = null)
+        }
+
+        if (null !== $address && $countryId = $address->id_country) {
+            return $countryId;
```
I dislike the affectation in the test. What do you think about

```php
        if (null !== $address && $address->id_country) {
            return $address->id_country;
        }
```

?

michaelKaefer replied:
@Quetzacoalt91 Thank you! I made another commit.

Quetzacoalt91 commented on classes/Tools.php:
```diff
@@ -611,22 +611,58 @@ public static function switchLanguage(Context $context = null)
+    {
+        if ($countryId = (int) Tools::getValue('id_country')) {
+            return $countryId;
+        }
```
It seems in the previous version, that case can be overridden if `Configuration::get('PS_DETECT_COUNTRY')` is true. In my opinion the behavior is changing here (and this explain why I was talking about tests)

mickaelandrieu replied:
@michaelKaefer if you don't know how to write tests with PHPUnit and you want to try, we can help you to help us 🤣

michaelKaefer committed:
Edited affectation

michaelKaefer committed:
Refactor ToolsCoreTest and add a test

michaelKaefer commented:
@Quetzacoalt91 @mickaelandrieu I commited a test. Yes, I definitly need your help. I can't even figure out what this method should do.. If it really works correctly in its old state, of course I can try to refactor it to produce the same results. Can you approve that the old method works correctly after having a look on the test results?

mickaelandrieu commented:
@michaelKaefer I didn't know this class was tested o_O 

I think we're good here then, thank you!

mickaelandrieu commented:
wdyt @Quetzacoalt91 ? at least;, it's tested now

PierreRambaud commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[tests/Unit/Classes/ToolsCoreTest.php, line 434 at r3](https://reviewable.io/reviews/prestashop/prestashop/9060#-LCgmucB5s_KrJ15Yo_V:-LCgmucC11McEGjysVUG:b-jlh4ww) ([raw file](https://github.com/prestashop/prestashop/blob/9a7c032954c98fd1839d462d2b9825493e821f63/tests/Unit/Classes/ToolsCoreTest.php#L434)):*
> ```PHP
> 
>         // Set dependencies
>         $address = new \Address();
> ```

Can you use namespace please for `Address`, `Configuration` and `Tools`?

---


*Comments from [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9060)*
<!-- Sent from Reviewable.io -->

michaelKaefer committed:
fix data provider; add use statements

michaelKaefer committed:
fix

michaelKaefer committed:
data provider

michaelKaefer commented:
@mickaelandrieu @Quetzacoalt91 The test only works if in your DB there is a country with id=21 and iso-code=EN. Otherwise the test is senseless. Is this guaranteed by the testing setup we use? So the refactor returns now exactly the same as before, I fixed the refactor. But the test is a problem if there is no such country as described. @PierreRambaud I added the requested use statements.

PierreRambaud commented:
@mickaelandrieu Should be ok, wdyt?

michaelKaefer commented:
@PierreRambaud @Progi1984 I'll close this. I think you got enough work to do.

matks commented:
> @PierreRambaud @Progi1984 I'll close this. I think you got enough work to do.

Thanks for the time you gave to this project 🤝
